### DISCUSSION TITLE
Help setting up a basic server

### DISCUSSION THREAD

**Discussion by JimmyTang999** [pymodbus-dev/pymodbus#2197] [Q&A] [answered]
here is what i have so far:

`
from pymodbus.datastore.context import ModbusSlaveContext, ModbusServerContext
from pymodbus.datastore.store import ModbusSequentialDataBlock
from pymodbus.server.async_io import StartTcpServer

store = ModbusSlaveContext(
    di=ModbusSequentialDataBlock(0, [0]*100),
    co=ModbusSequentialDataBlock(0, [0]*100),
    hr=ModbusSequentialDataBlock(0, [0]*100),
    ir=ModbusSequentialDataBlock(0, [0]*100))
context = ModbusServerContext(slaves=store, single=True)

server = StartTcpServer(context, address=("localhost", 5020))
`
and here is what it says

`
StartTcpServer() takes 0 positional arguments but 1 was given
`

spent some time aggravating chatgpt but still couldn't get it

**Comment by janiversen** [comment, marked as answer]
please see our examples and api definition in the documentation. you are not calling starttcpserver correctly.

**Comment by JimmyTang999** [comment]
when i copy paste directly from the example from the documentation and run, i try to connect with radzio and get message timeouts. 

`
from pymodbus.server import ModbusSimulatorServer

async def run():
    simulator = ModbusSimulatorServer(
        modbus_server="my server",
        modbus_device="my device",
        http_host="127.0.0.1",
        http_port=8080)
    await simulator.run_forever(only_start=True)
    ...
    await simulator.stop()
`

This is with firewall disabled
![ModbusIssue](https://github.com/pymodbus-dev/pymodbus/assets/160504910/6fcb3643-b294-4770-8aa2-41655b0864d2)

**Reply by janiversen** [reply]
please use the debug log call, so we can see what is happening...

Did you remember to change the configuration to match your network?

**Reply by JimmyTang999** [reply]
a little more info- i am running everything on the same computer. the pymodbus server as well as the radzio  software. 
not even far enough along here to try something over a network (hence the 127.0.0.1 localhost address). 

logging isn't working

heres what i did

`
import logging
import logging.handlers as Handlers
from pymodbus.server import ModbusSimulatorServer

logging.basicConfig()
log = logging.getLogger()
log.setLevel(logging.DEBUG)

logging.getLogger('pymodbus').setLevel(logging.DEBUG)

async def run():
    simulator = ModbusSimulatorServer(
        modbus_server="my server",
        modbus_device="my device",
        http_host="127.0.0.1",
        http_port=8080)
    await simulator.run_forever(only_start=True)
    await simulator.stop()
`

when you say "change the configuration" are you alluding to the json file found here: 
PyModbus\Lib\site-packages\pymodbus\server\simulator
?

**Reply by janiversen** [reply]
that does work, we. have a call in the API.

It is in most of our examples and in the documentation

**Reply by janiversen** [reply]
the json file is the configuration....but do you want a simulator or just a simple server.

The simulator is a lot more complicated to handle than an ordinary simple server. There are examples for both use cases.

**Reply by janiversen** [reply]
I think you will benefit from reading the simulation part of our documentation, it explains in detail how to configure and use the simulator.

**Reply by JimmyTang999** [reply]
tryig to figure out where to set DeviceID for the server - by this i mean an integer from 1-247. I realize this is for a TCP modbus server but i would like to know if it is possible to set the DeviceID of the server. 
![WhereToSetDeviceID](https://github.com/pymodbus-dev/pymodbus/assets/160504910/2716e5c4-e7d9-4cc6-9ad0-d7bc6b18a4de)

**Reply by janiversen** [reply]
A server have no deviceId but it responds when adressing device in the server, that is called slave which is the same as deviceId. This is supported by both simulator and server.

A server have a device identifier, which is a string, there is a parameter to set that.

**Reply by janiversen** [reply]
Correction the simulator in the repo, do not support multiple deviveIDs it responds to all.

**Reply by JimmyTang999** [reply]
finally got it- [this](https://www.youtube.com/watch?v=5qMd0KBIC3E) was helpful as were the answers above. Thank you
